API 的插件框架

Cirn09 opened the pull request:
相关：https://github.com/qiandao-today/qiandao/issues/354


现在这个是第三版设计，废弃的第二版中已经实现了几个 API
https://github.com/Cirn09/qiandao/tree/api-v2

Cirn09 committed:
装饰器风格的 API plugin 框架

Cirn09 committed:
API examples

Cirn09 committed:
API plugin 框架更新

Cirn09 committed:
api://api/xx -> api://v1/xx

a76yyyy committed:
Update Error message display

a76yyyy committed:
Fix bug from codeQL

Cirn09 commented on web/handlers/api/example.py:
```diff
@@ -42,15 +44,15 @@
 class Echo1(ApiBase):
     api_name = "回声"
-    api_description = "输出 = 输入"  # 支持 HTML 标签，比如 <br/>
+    api_description = "输出 = 输入"  # 不支持 HTML 标签，比如 <br/>
```
我觉得在这里支持 HTML 标签是有必要的，用来加粗标红注意事项，给出外部链接之类的，比如：
https://github.com/Cirn09/qiandao/blob/4042fcaf896e4e408378ff703349d8f2f3feeeba/web/handlers/api/time.py#L47-L53
而且 api_description 不是普通用户可控的，不用担心注入攻击

a76yyyy replied:
这里可以，我的意思是入参那里不行，我注释没表述好

Cirn09 commented on web/handlers/api/__init__.py:
```diff
@@ -0,0 +1,317 @@
+        if escape_html:
+            data = escape(json.dumps(data, ensure_ascii=ensure_ascii, indent=indent), quote=escape_quote)
+        else:
+            data = json.dumps(data, ensure_ascii=ensure_ascii, indent=indent).replace('</', '<\\/')
```
这里应该没必要进行转义，因为 Content-Type 已经设置为 `application/json` 了

更进一步，API 的返回内容默认为 `text/plain` 吧，避免 API 作者忘记转义（比如我）；`bytes` 类型则 `application/octet-stream`；JSON `application/json`。
应该没有 API 需要使用其他类型的需求，所以就不用允许 API 自己设置 Content-Type 了

a76yyyy replied:
我的建议是默认返回JSON格式的`code，message，data`信息

a76yyyy replied:
既然是api就一步到位，规范一下

Cirn09 replied:
这样设计的话，我觉得对应的系统对 JSON respond 的处理能力需要增强，目前系统对返回内容好像只支持正则匹配。

思考了一下好像不难实现，提供一个 respond content 的魔法变量和一个 JSON 函数应该就可以了，比如 `{{ json_parse(content)['code'] }}`

Cirn09 replied:
> 这样设计的话，我觉得对应的系统对 JSON respond 的处理能力需要增强，目前系统对返回内容好像只支持正则匹配。
> 
> 思考了一下好像不难实现，提供一个 respond content 的魔法变量和一个 JSON 函数应该就可以了，比如 `{{ json_parse(content)['code'] }}`


现在的模式：
![image](https://github.com/qd-today/qd/assets/25722111/6ca5e641-29c0-4e7c-857e-698f36e40a18)

新UI： 类型(下拉: 简单, 正则(和旧版兼容), 代码), 来源(下拉: respond.[content, header, status]), 变量名, 规则, 结果
类型为代码时则不显示来源，可以使用 `{{ respond.content }}` 之类的

a76yyyy replied:
没有看到新UI的界面?

Cirn09 replied:
在写了.jpg，先把方案写下来讨论一下

Cirn09 committed:
强制 JSON 的 API 框架

Cirn09 committed:
BodyArgument 自动解码

a76yyyy commented on web/handlers/api/__init__.py:
```diff
@@ -141,20 +141,41 @@ class BodyArgument(ArgumentBase):
 
         return super().__post_init__()
 
+    def try_decode(self, api: "ApiBase") -> str | bytes:
```
可以参考
https://github.com/qd-today/qd/blob/master/libs/utils.py#L434

https://github.com/qd-today/qd/blob/master/libs/utils.py#L466

a76yyyy committed:
Merge branch 'master' into api-v3

Signed-off-by: a76yyyy <56478790+a76yyyy@users.noreply.github.com>

Cirn09 committed:
更新 BodyArgument 的自动解码

Cirn09 committed:
在 jinja 中加入 json 相关函数

Cirn09 committed:
Merge branch 'api-v3' of https://github.com/Cirn09/qiandao into api-v3

Cirn09 committed:
合并时的遗留问题

rename BaseWebSocket -> BaseWebSocketHandler
rename BaseHandler -> _BaseHandler
remove about.html {% raw %}

a76yyyy committed:
Merge branch 'master' into api-v3

Signed-off-by: a76yyyy <56478790+a76yyyy@users.noreply.github.com>
